### PR TITLE
Add historical coverage script for bb

### PR DESCRIPTION
**Pull request by tapishr**
- [x] Running end to end
- [x] Reverse engineers git blame from bitbucket apis
- [x] Rest of the script is similar to github with data format changes

Output - 
```
Workspace: alokit-innovations
  Repository: dev-profiler
        Processing Pr #10
     Processing PR #10 with author: vibi-test and reviewers: ['Tapish Rathore', 'vibi-test']
    Total deletions considered in PR: 12, unapproved deletions: 0, repo_coverage: 1, prs_considered: 1
    ============<Coverage percentage for Merged PR #10: 100.00%>=============
        Processing Pr #4
        No non-author deletions in PR #4, not included in coverage calculation
        Processing Pr #2
        No non-author deletions in PR #2, not included in coverage calculation
        Processing Pr #1
        No non-author deletions in PR #1, not included in coverage calculation
    # of prs considered = 1 in repo dev-profiler. workspace_coverage: 1.0
    ~~~~~~~~~~~~><><><><><><Coverage percentage for Repository dev-profiler = 100.00%><>><><><><~~~~~~~~~~~~~~
  Repository: repo-profiler-pipe
        Processing Pr #7
        No non-author deletions in PR #7, not included in coverage calculation
        Processing Pr #6
    ============<Relevant authors who did not review the PR: {'Tapish Rathore'}>=============
     Processing PR #6 with author: Avikalp Gupta and reviewers: ['Avikalp Gupta']
    Total deletions considered in PR: 3, unapproved deletions: 3, repo_coverage: 0.0, prs_considered: 1
    ============<Coverage percentage for Merged PR #6: 0.00%>=============
        Processing Pr #5
        No non-author deletions in PR #5, not included in coverage calculation
        Processing Pr #4
        No non-author deletions in PR #4, not included in coverage calculation
        Processing Pr #3
            Skipped Users = {'{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
        No non-author deletions in PR #3, not included in coverage calculation
        Processing Pr #2
            Skipped Users = {'{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
        No non-author deletions in PR #2, not included in coverage calculation
        Processing Pr #1
            Skipped Users = {'{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
            Skipped Users = {'{"raw": "Bitbucket Pipelines Push Bot <commits-noreply@bitbucket.org>", "type": "author"}', '{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
            Skipped Users = {'{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
            Skipped Users = {'{"raw": "Bitbucket Pipelines Push Bot <commits-noreply@bitbucket.org>", "type": "author"}', '{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
            Skipped Users = {'{"raw": "bitbucket-pipelines <commits-noreply@bitbucket.org>", "type": "author"}'}
        No non-author deletions in PR #1, not included in coverage calculation
    # of prs considered = 1 in repo repo-profiler-pipe. workspace_coverage: 1.0
    ~~~~~~~~~~~~><><><><><><Coverage percentage for Repository repo-profiler-pipe = 0.00%><>><><><><~~~~~~~~~~~~~~
  Repository: pipeline-testing
No valid prs in repo pipeline-testing, not including in coverage calculation
  Repository: pipeline
No valid prs in repo pipeline, not including in coverage calculation
  Repository: demo-repo
        Processing Pr #3
        No non-author deletions in PR #3, not included in coverage calculation
        Processing Pr #2
        No non-author deletions in PR #2, not included in coverage calculation
No valid prs in repo demo-repo, not including in coverage calculation
Total repos considered = 2, total_coverage = 0.5
######################### Coverage percentage for Workspace alokit-innovations = 50.00% #####################################





Workspace: alokit-innovations-test
  Repository: dev-profiler
No valid prs in repo dev-profiler, not including in coverage calculation
  Repository: dev-profiler-1
No valid prs in repo dev-profiler-1, not including in coverage calculation
  Repository: dev-profiler-2
No valid prs in repo dev-profiler-2, not including in coverage calculation
  Repository: dev-profiler-3
No valid prs in repo dev-profiler-3, not including in coverage calculation
  Repository: dev-profiler-4
No valid prs in repo dev-profiler-4, not including in coverage calculation
  Repository: dev-profiler-5
No valid prs in repo dev-profiler-5, not including in coverage calculation
  Repository: dev-profiler-6
No valid prs in repo dev-profiler-6, not including in coverage calculation
  Repository: dev-profiler-7
No valid prs in repo dev-profiler-7, not including in coverage calculation
  Repository: dev-profiler-8
No valid prs in repo dev-profiler-8, not including in coverage calculation
  Repository: dev-profiler-9
No valid prs in repo dev-profiler-9, not including in coverage calculation
No valid repos in workspace alokit-innovations-test, not including in coverage calculation
Workspaces analyzed: 1
Total coverage across all workspaces: 50.00
```

### CODE REVIEW

File: coverage/historical/bitbucket/historical_coverage_report.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        return (total_unapproved_deletions, total_deletions, total_unassigned)
+    old_commit_hash = pr['source']['commit']['hash']
+    new_commit_hash = pr['destination']['commit']['hash']
+    filepath = file['old']['path']
+    diff = get_diff(workspace, repo['slug'], old_commit_hash, new_commit_hash, filepath)
+    # blame = get_blame(workspace, repo, old_commit_hash, filepath)
+    # print(f"\n\n\n\n pr = {pr}")
+    # if not blame:
+    #     print(f"Could not get blame for commit: {pr['baseRefOid']}, file: {file['filename']}")
+    #     return (total_unapproved_deletions, total_deletions, total_unassigned)
```
Comment (avikalpg) [inline coverage/historical/bitbucket/historical_coverage_report.py]: phaltu

Comment (avikalpg) [inline coverage/historical/bitbucket/historical_coverage_report.py]:
```suggestion
```

File: coverage/historical/bitbucket/historical_coverage_report.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+    diff_cache[cache_key] = diff
+    return diff
+
+def get_commit_author(commit_url, commit_cache):
+    """
+    Get commit author
+    """
+    headers = {
+        "Authorization": f"Basic {base64_encoded_credentials}"
+    }
```
Comment (avikalpg) [inline coverage/historical/bitbucket/historical_coverage_report.py:411]: You have added this Authorization header each time you initiate a request. This is not required and you just do this once in that whole file.

**Review by avikalpg — APPROVED** [review]
Only low LOGAF changes suggested.

**Review by MuskanPaliwal — APPROVED** [review]

**Pull request merged**